src/dos/drive_fat.cpp

```diff
@@ -736,11 +738,14 @@ fatDrive::fatDrive(const char *sysFilename,
 		imgDTAPtr = RealMake(imgDTASeg, 0);
 		imgDTA    = new DOS_DTA(imgDTAPtr);
 	}
-
-	diskfile = fopen_wrap(sysFilename, "rb+");
+	diskfile = roflag ? NULL : fopen_wrap(sysFilename, "rb+");
 	if (!diskfile) {
-		created_successfully = false;
-		return;
+		diskfile = fopen_wrap(sysFilename, "rb");
+		if (!diskfile) {
+			created_successfully = false;
+			return;
+		} else
+			readonly = true;
```

Comment:
Noticed three areas that repeat the same code pattern. 
I think we can use one helper function:

```c++

// A helper for fopen_wrap that will fallback to read-only if read-write isn't possible.
// In the fallback case, is_readonly arguement is toggled to true.
// In all cases, a pointer to the file is returned (or nullptr on failure).
FILE *fopen_wrap_ro_fallback(const std::string &filename, bool &is_readonly)
{
	// Try with the requested permissions
	const auto permissions = is_readonly ? "rb" : "rb+";
	FILE *fp = fopen_wrap(filename.c_str(), permissions);

	// Fallback to read-only?
	const bool failed_opening_rw = (!fp && !is_readonly);
	if (failed_opening_rw) {
		fp = fopen_wrap(filename.c_str(), "rb");
		if (fp) {
			is_readonly = true;
			LOG_INFO("FILESYSTEM: Opened %s read-only per host-filesystem permissions",
			         filename.c_str());
		}
	}
	// Note: if failed, the caller should provide a context-specific failure message
	return f;
}
```

And then call that in all three areas?

Reply:
Thanks for the suggestion of using the helper function, which reduces duplicates.